src/bin/tikv-ctl.rs

```diff
@@ -21,25 +21,387 @@ extern crate clap;
+        (None, None) => {
+            eprintln!(r#"please pass "host" or "db""#);
+            process::exit(1);
+        }
+        (None, Some(kv_path)) => {
+            let db = util::rocksdb::open(kv_path, ALL_CFS).unwrap();
+            let raft_db = if let Some(raft_path) = raft_db {
+                util::rocksdb::open(raft_path, &[CF_DEFAULT]).unwrap()
+            } else {
+                let raft_path = PathBuf::from(kv_path).join("../raft");
+                util::rocksdb::open(raft_path.to_str().unwrap(), &[CF_DEFAULT]).unwrap()
+            };
+            Box::new(Debugger::new(Engines::new(Arc::new(db), Arc::new(raft_db)))) as
+                Box<DebugExecutor>
+        }
+        (Some(remote), None) => {
+            let env = Arc::new(Environment::new(1));
+            let channel = ChannelBuilder::new(env).connect(remote);
+            let client = DebugClient::new(channel);
+            Box::new(client) as Box<DebugExecutor>
+        }
+    }
+}
+
+trait DebugExecutor {
+    fn dump_key_value(&self, cf: &str, key: Vec<u8>) {
+        let value = self.get_value_by_key(cf, key);
+        println!("value: {}", escape(&value));
+    }
+
+    fn dump_region_size(&self, region: u64, cfs: Vec<&str>) {
+        let sizes = self.get_region_size(region, cfs);
+        if sizes.len() > 1 {
+            let total_size = sizes.iter().map(|t| t.1).sum::<usize>() as u64;
+            println!("total region number: {}", sizes.len());
+            println!("total region size: {}", convert_gbmb(total_size));
+        }
+        for (cf, size) in sizes {
+            println!("cf: {}", cf);
+            println!("region size: {}", convert_gbmb(size as u64));
+        }
+    }
+
+    fn dump_all_region_size(&self, cfs: Vec<&str>) {
+        for region in self.get_all_meta_regions() {
+            self.dump_region_size(region, cfs.clone());
+        }
+    }
+
+    fn dump_region_info(&self, region: u64, skip_tombstone: bool) {
+        let r = self.get_region_info(region);
+        if skip_tombstone {
+            let region_state = r.region_local_state.as_ref();
+            if region_state.map_or(false, |s| s.get_state() == PeerState::Tombstone) {
+                return;
+            }
+        }
+        let region_state_key = keys::region_state_key(region);
+        let raft_state_key = keys::raft_state_key(region);
+        let apply_state_key = keys::apply_state_key(region);
+        println!("region state key: {}", escape(&region_state_key));
+        println!("region state: {:?}", r.region_local_state);
+        println!("raft state key: {}", escape(&raft_state_key));
+        println!("raft state: {:?}", r.raft_local_state);
+        println!("apply state key: {}", escape(&apply_state_key));
+        println!("apply state: {:?}", r.raft_apply_state);
+    }
+
+    fn dump_all_region_info(&self, skip_tombstone: bool) {
+        for region in self.get_all_meta_regions() {
+            self.dump_region_info(region, skip_tombstone);
```

Comment:
I think you need to print region id before display its information.

Reply:
fixed.